tools/ci_connector_ops/ci_connector_ops/pipelines/actions/environments.py

```diff
@@ -197,3 +202,200 @@ async def with_ci_connector_ops(context: ConnectorTestContext) -> Container:
+    Args:
+        context (ConnectorTestContext): The current connector test context.
+        sources_to_include (List[str], optional): List of additional source path to mount to the container. Defaults to None.
+        docker_cache_volume_name (Optional[str], optional): The name of the docker cache volume. Changing the volume name provides docker context isolation. Defaults to None.
+
+    Returns:
+        Container: A container with Gradle installed and Java sources from the repository.
+    """
+    airbyte_gradle_cache: CacheVolume = context.dagger_client.cache_volume(f"{context.connector.technical_name}_airbyte_gradle_cache")
+    root_gradle_cache: CacheVolume = context.dagger_client.cache_volume(f"{context.connector.technical_name}_root_gradle_cache")
+
+    include = [
+        ".env",
+        "build.gradle",
+        "deps.toml",
+        "gradle.properties",
+        "gradle",
+        "gradlew",
+        "LICENSE_SHORT",
+        "publish-repositories.gradle",
+        "settings.gradle",
+        "build.gradle",
+        "tools/gradle",
+        "spotbugs-exclude-filter-file.xml",
+        "airbyte-api",
+        "airbyte-commons-cli",
+        "airbyte-commons-protocol",
+        "airbyte-commons",
+        "airbyte-config",
+        "airbyte-connector-test-harnesses",
+        "airbyte-db",
+        "airbyte-integrations/bases",
+        "airbyte-integrations/connectors/source-jdbc",
+        "airbyte-integrations/connectors/source-relational-db",
+        "airbyte-json-validation",
+        "airbyte-protocol",
+        "airbyte-test-utils",
+        "buildSrc",
+        "tools/bin/build_image.sh",
+        "tools/lib/lib.sh",
+    ]
+
+    python_install_dependencies = [
+        "build-essential",
+        "zlib1g-dev",
+        "libncurses5-dev",
+        "libgdbm-dev",
+        "libnss3-dev",
+        "libssl-dev",
+        "libreadline-dev",
+        "libffi-dev",
+        "libsqlite3-dev",
+        "wget",
+        "libbz2-dev",
+    ]
+
+    if sources_to_include:
+        include += sources_to_include
+
+    shared_tmp_volume = ("/tmp", context.dagger_client.cache_volume("share-tmp-gradle"))
+    openjdk_with_docker = (
+        context.dagger_client.container()
+        # Use openjdk image because it's based on Debian. Alpine with Gradle and Python causes filesystem crash.
+        .from_("openjdk:17.0.1-jdk-slim")
+        .with_exec(["apt-get", "update"])
+        .with_exec(["apt-get", "install", "-y", "curl"])
+        .with_env_variable("VERSION", "23.0.1")
+        .with_exec(["sh", "-c", "curl -fsSL https://get.docker.com | sh"])
+        .with_exec(["apt-get", "install", "-y"] + python_install_dependencies)
+        .with_exec(
+            [
+                "apt-get",
+                "install",
+                "-y",
+                "python3",
+                "python3-pip",
+            ]
+        )
+        # TODO it seems that the Gradle Python plugin can't create virtualenv
```

Comment:
@alafanechere Whats the error were seeing? What do you think is happening? Im sure we can fix this..... but there might be alot of pain involved (build, fail, change, build, fail, cry, change...)

Reply:
Edit: I found the root cause.
My host `.venv` was mounted to the container, causing a corrupted python install.

Reply:
🥳

Reply:
Glad you made it work 🙂, but let me just chime in to clarify that you don't need to use `virtualenv` or call `./bin/activate` (not sure how you're activating here).

The most elegant way to activate a Python virtual environment in a container is to just set `VIRTUAL_ENV` to the path where your virtual env dir is, then update `PATH` to add it's `$VIRTUAL_ENV/bin` directory.

Example here:

https://github.com/dagger/dagger/blob/a04e61f797628fad0e471e0644b551f96053cfc0/internal/mage/sdk/python.go#L191-L196

And more explanation at [Elegantly activating a virtualenv in a Dockerfile](https://pythonspeed.com/articles/activate-virtualenv-dockerfile/).

Reply:
Glad you made it work 🙂, let me chime in just to clarify that you don't need to use `virtualenv` in a container. The most elegant way to activate a Python virtual environment in a container is to set `VIRTUAL_ENV` to the path where your virtual env is, then add `$VIRTUAL_ENV/bin` to `PATH`.

You can see an [example](https://github.com/dagger/dagger/blob/a04e61f797628fad0e471e0644b551f96053cfc0/internal/mage/sdk/python.go#L191-L196) in our own CI, and an explanation in [Elegantly activating a virtualenv in a Dockerfile](https://pythonspeed.com/articles/activate-virtualenv-dockerfile/).

Reply:
@helderco yes I'd prefer to not use venv in the containers. But in this context I want to run a gradle task that rely on a python plugin which is using virtual envs. To keep the gradle files untouched I'd prefer to adapt the dagger pipeline to the gradle environment.

Reply:
You misunderstood. I'm just saying how to activate a virtual env in containers, not to avoid using them. When I said you don't need `virtualenv`, I'm referring to the pypi package as you're installing it with pip.

Reply:
If you look at the example it should become clearer.